Ingress/controllers/nginx-third-party/controller.go

```diff
@@ -0,0 +1,237 @@
+	"text/template"
+
+	flag "github.com/spf13/pflag"
+
+	"github.com/fatih/structs"
+	"github.com/golang/glog"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/apis/extensions"
+	"k8s.io/kubernetes/pkg/client/unversioned"
+	"k8s.io/kubernetes/pkg/fields"
+	"k8s.io/kubernetes/pkg/labels"
+	"k8s.io/kubernetes/pkg/util"
+)
+
+const (
+	healthPort = 10249
+)
+
+var (
+	flags = flag.NewFlagSet("", flag.ContinueOnError)
+
+	defaultSvc = flags.String("default-backend-service", "kube-system/default-http-backend",
+		`Service used to serve a 404 page for the default backend. Takes the form
+		namespace/name. The controller uses the first node port of this Service for
+		the default backend.`)
+
+	customErrorSvc = flags.String("custom-error-service", "kube-system/custom-error-backend",
+		`Service used that will receive the errors from nginx and serve a custom error page. 
+		Takes the form namespace/name. The controller uses the first node port of this Service 
+		for the backend.`)
+
+	// If you have pure tcp services or https services that need L3 routing, you
+	// must specify them by name. Note that you are responsible for:
+	// 1. Making sure there is no collision between the service ports of these services.
+	//	- You can have multiple <mysql svc name>:3306 specifications in this map, and as
+	//	  long as the service ports of your mysql service don't clash, you'll get
+	//	  loadbalancing for each one.
+	// 2. Exposing the service ports as node ports on a pod.
+	// 3. Adding firewall rules so these ports can ingress traffic.
+
+	tcpServices = flags.String("tcp-services", "", `Comma separated list of tcp/https
+		namespace/serviceName:portToExport pairings. This assumes you've opened up the right
+		hostPorts for each service that serves ingress traffic. Te value of portToExport indicates the 
+		port to listen inside nginx, not the port of the service.`)
+
+	configurationName = flags.String("configuration-name", "", `Name of the NginxIngressLoadBalancer 
+		third party resource that contains the custom configuration to be used in nginx.
+		Takes the form namespace/name.`)
+
+	healthzPort = flags.Int("healthz-port", healthPort, "port for healthz endpoint.")
+)
+
+type k8sService struct {
+	ServiceName string
+	ServicePort string
+	Namespace   string
+	ExposedPort string
+}
+
+func registerHandlers() {
+	http.HandleFunc("/healthz", func(w http.ResponseWriter, r *http.Request) {
+		w.WriteHeader(200)
+		w.Write([]byte("ok"))
+	})
+
+	glog.Fatalf(fmt.Sprintf("%v", http.ListenAndServe(fmt.Sprintf(":%v", *healthzPort), nil)))
+}
+
+func main() {
+	pflagGlog(flags)
+
+	flags.Parse(os.Args)
+
+	if *defaultSvc == "" {
+		glog.Fatalf("Please specify --default-backend")
+	}
+
+	var ingClient unversioned.IngressInterface
+
+	kubeClient, err := unversioned.NewInCluster()
+	if err != nil {
+		glog.Fatalf("failed to create client: %v", err)
+	} else {
+		ingClient = kubeClient.Extensions().Ingress(api.NamespaceAll)
+	}
+
+	funcMap := template.FuncMap{
```

Comment:
be careful, putting more template functions decreases the debuggability of the controller. If possible we should minimize this. Also, can we move the templating out of the controller (or into a custom type that hides the details from the controller)?